### PR TITLE
Create more robust toolchain control with error recording for repeatable testing

### PR DESCRIPTION
**Pull request by jimidle**
Here, we extend the basic toolchain we had such that all errors are captured and errors at any stage in the workflow `parse->visit->optimize->transpile` prevent an attempt to continue the workflow. This prevents, for example, attempts to generate IR from a parse tree generated with syntax errors. There is no point trying to generate IR from a broken ParseTree.

We also provide some basic source code providers for directories and files. 

Finally, we provide a command line interface for transpilation such that developers can look at the results of any particular input. This includes the ability to retrieve query history (Snowflake only for now), attempt to compile each retrieved query and record the results as jsonl output. The code the command line uses can be the basis for an ongoing integration test further down the line.

### CODE REVIEW

**Review by nfx — CHANGES_REQUESTED** [review]
use PlanParser abstractions that are already there

File: core/src/main/scala/com/databricks/labs/remorph/transpilers/SnowflakeToDatabricksTranspiler.scala [hunk 1]
```diff
@@ -21,21 +26,81 @@ class SnowflakeToDatabricksTranspiler extends BaseTranspiler {
...
+      Result(stage = WorkflowStage.PARSE, tree = Some(tree))
+    }
+  }
+
+  implicit val formats: Formats = Serialization.formats(NoTypeHints)
+
+  override def visit(tree: Result): Result = {
+    try {
+      val plan = astBuilder.visit(tree.tree.get)
+      Result(stage = WorkflowStage.PLAN, plan = Some(plan), tree = tree.tree)
```
[the remark this answers is not in the document]
Comment (jimidle) [reply core/src/main/scala/com/databricks/labs/remorph/transpilers/SnowflakeToDatabricksTranspiler.scala]: Changed to use Monads

File: core/src/main/scala/com/databricks/labs/remorph/transpilers/SnowflakeToDatabricksTranspiler.scala [hunk 2]
```diff
@@ -21,21 +26,81 @@ class SnowflakeToDatabricksTranspiler extends BaseTranspiler {
...
     val parser = new SnowflakeParser(tokenStream)
-    val errHandler = new ProductionErrorCollector(input, "-- test string --")
+    val errListener = new ProductionErrorCollector(input.source, input.filename)
     parser.removeErrorListeners()
-    parser.addErrorListener(errHandler)
+    parser.addErrorListener(errListener)
     val tree = parser.snowflakeFile()
-    astBuilder.visit(tree)
+    if (errListener.errorCount > 0) {
+      Result(stage = WorkflowStage.PARSE, errorsJson = errListener.errorsAsJson, tree = Some(tree))
```
[the remark this answers is not in the document]
Comment (jimidle) [reply core/src/main/scala/com/databricks/labs/remorph/transpilers/SnowflakeToDatabricksTranspiler.scala]: Changed to use Monads

**Review by nfx — COMMENTED** [review]
i'll issue follow-up refactorings

**Pull request merged**